### PR TITLE
Evaluation tests

### LINKED ISSUE #13: Add evaluation.spec.js tests
Based on this tag https://github.com/kblok/playwright/tree/first-snapshot. We don't need to get the tests green. We only need to incoporate the code. Let's also check if we can get code from Puppeteer

### PR DESCRIPTION
**Pull request by Meir017**
closes #13

### CODE REVIEW

File: src/PlaywrightSharp.Abstractions/IJSHandle.cs [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
     {
+        /// <summary>
+        /// Returns a JSON representation of the object.
+        /// </summary>
+        /// <returns>Task.</returns>
+        /// <remarks>
+        /// The method will return an empty JSON if the referenced object is not stringifiable. It will throw an error if the object has circular references.
+        /// <returns>A <see cref="Task"/> that completes when the evaluation is completed, yielding an <see cref="object"/> with the json value of the handle.</returns>
+        /// </remarks>
+        Task<object> JsonValueAsync();
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IJSHandle.cs]:
```
        Task<T> GetJsonValueAsync<T>();
```

File: src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Frame.evaluate</playwright-describe>
+        ///<playwright-it>should dispose context on navigation</playwright-it>
+        [Fact]
+        public async Task ShouldDisposeContextOnNavigation()
+        {
+            await Page.GoToAsync(TestConstants.ServerUrl + "/frames/one-frame.html");
+            Assert.Equal(2, Page.Frames.Length);
+            // TODO: expect(page._delegate._contextIdToContext.size).toBe(4);
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs]: _delegate => `Delegate` can be an internal `IPageDelegate` property.
_contextIdToContext => ContextIdToContext and internal `Dictionary<int, IExecutionContext>`

File: src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            var mainFrame = Page.MainFrame;
+            Assert.Contains("localhost", await mainFrame.EvaluateAsync<string>("() => window.location.href"));
+            await Page.GoToAsync(TestConstants.CrossProcessUrl + "/empty.html");
+            Assert.Contains("127", await mainFrame.EvaluateAsync<string>("() => window.location.href"));
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Frame.evaluate</playwright-describe>
+        ///<playwright-it>should allow cross-frame js handles</playwright-it>
+        [Fact(Skip = "Not implemented")]
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs]:
```suggestion
        [Fact(Skip = "Skipped in Playwright")]

```

File: src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            await Page.GoToAsync(TestConstants.ServerUrl + "/frames/one-frame.html");
+            var bodyHandle = await Page.MainFrame.ChildFrames[0].QuerySelectorAsync("body");
+            var result = await Page.EvaluateAsync<string>("body => body.innerHTML", bodyHandle);
+            Assert.Equal("<div>Hi, I\'m frame</div>", result.Trim());
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Frame.evaluate</playwright-describe>
+        ///<playwright-it>should not allow cross-frame element handles when frames do not script each other</playwright-it>
+        [Fact(Skip = "Not implemented")]
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs]: Skip only in Firefox

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+            Assert.Null(await Page.EvaluateAsync<object>(@"() => {
+                return { foo: [{ a: Symbol('foo4') }] };
+            }"));
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should work with function shorthands</playwright-it>
+        [Fact(Skip = "Not relevant for C#, js specific")]
+        public Task ShouldWorkWithFunctionShorthands()
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]:
```suggestion
        public void ShouldWorkWithFunctionShorthands() {}

```

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+            return Task.CompletedTask;
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should work with unicode chars</playwright-it>
+        [Fact]
+        public async Task ShouldWorkWithUnicodeChars()
+        {
+            var result = await Page.EvaluateAsync<int>("a => a['Σ╕¡µצחσ¡קτ¼ª'], { 'Σ╕¡µצחσ¡קτ¼ª': 42}");
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Something when wrong here. Check this out https://github.com/kblok/puppeteer-sharp/blob/7fca5eeeaa9e138af09e24bd98d11d128cac260b/lib/PuppeteerSharp.Tests/PageTests/EvaluateTests.cs#L80

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+        {
+            var result = await Page.EvaluateAsync<bool>("(a, b) => Object.is (a, undefined) && Object.is (b, 'foo')", null, "foo");
+            Assert.True(result);
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should properly serialize undefined arguments</playwright-it>
+        [Fact(Skip = "Not relevant for C#, js specific")]
+        public Task ShouldProperlySerializeUndefinedArguments()
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]:
```suggestion
        public void ShouldProperlySerializeUndefinedArguments() {}

```

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+            return Task.CompletedTask;
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should properly serialize undefined fields</playwright-it>
+        [Fact]
+        public async Task ShouldProperlySerializeUndefinedFields()
+        {
+            Assert.Equal(JsonDocument.Parse("{}"), await Page.EvaluateAsync<JsonDocument>("() => ({ a: undefined})"));
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Let's use arrows in this class when we can.

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+                throw new Error(errorText);
+            }", exception.Message);
+            Assert.Contains(error.GetProperty("message").GetString(), exception.Message);
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should accept a string</playwright-it>
+        [Fact(Skip = "Not relevant for C#, js specific")]
+        public Task ShouldAcceptAString()
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]:
```suggestion
        public void ShouldAcceptAString() {}

```

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+        {
+            return Task.CompletedTask;
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should accept a string with semi colons</playwright-it>
+        [Fact(Skip = "Not relevant for C#, js specific")]
+        public Task ShouldAcceptAStringWithSemiColons()
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]:
```suggestion
        public void ShouldAcceptAStringWithSemiColons() {}

```

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+                window.location = '/empty.html';
+                return [42];
+            }");
+            Assert.Equal(new[] { 42 }, result);
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should transfer 100Mb of data from page to node.js</playwright-it>
+        [Fact(Skip = "Not implemented")]
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Skip only for firefox

File: src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs (hunk 2 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/Evaluation/FrameEvaluateTests.cs]: I just realized we can't do this with our interfaces, can't we do an Assert.Fail so we don't forget about these To-Dos

File: src/PlaywrightSharp.Abstractions/IJSHandle.cs [hunk 12]
```diff
@@ -1,9 +1,26 @@
...
     {
+        /// <summary>
+        /// Returns a JSON representation of the object.
+        /// </summary>
+        /// <returns>Task.</returns>
+        /// <remarks>
+        /// The method will return an empty JSON if the referenced object is not stringifiable. It will throw an error if the object has circular references.
+        /// <returns>A <see cref="Task"/> that completes when the evaluation is completed, yielding an <see cref="object"/> with the json value of the handle.</returns>
+        /// </remarks>
+        Task<object> GetJsonValueAsync();
```
Comment (kblok) [inline src/PlaywrightSharp.Abstractions/IJSHandle.cs]:
```
        Task<T> GetJsonValueAsync<T>();
```

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs (hunk 8 above)
Comment (kblok) [reply src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: @Meir017 ☝️

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+        {
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should properly serialize undefined fields</playwright-it>
+        [Fact]
+        public async Task ShouldProperlySerializeUndefinedFields()
+        {
+            Assert.Equal(new { a = (object)null }, await Page.EvaluateAsync<object>("() => ({ a: undefined})"));
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Arrows

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should properly serialize null arguments</playwright-it>
+        [Fact]
+        public async Task ShouldProperlySerializeNullArguments()
+        {
+            Assert.Null(await Page.EvaluateAsync<JsonDocument>("x => x", null));
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Arrows

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should properly serialize null fields</playwright-it>
+        [Fact]
+        public async Task ShouldProperlySerializeNullFields()
+        {
+            Assert.Equal(new { a = (object)null }, await Page.EvaluateAsync<object>("() => ({ a: null}))"));
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Arrows

File: src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+        }
+
+        ///<playwright-file>evaluation.spec.js</playwright-file>
+        ///<playwright-describe>Page.evaluate</playwright-describe>
+        ///<playwright-it>should return undefined for non-serializable objects</playwright-it>
+        [Fact]
+        public async Task ShouldReturnUndefinedForNonSerializableObjects()
+        {
+            Assert.Null(await Page.EvaluateAsync<object>("() => window"));
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: Arrows

Comment (Meir017) [reply src/PlaywrightSharp.Tests/Evaluation/PageEvaluateTests.cs]: we should enforce this style ;)

**Pull request merged**